GitCommands/Utils/EnumHelper.cs

```diff
@@ -16,9 +16,25 @@ public static class EnumHelper
             return attributes.FirstOrDefault()?.Description ?? value.ToString();
         }
 
-        public static T[] GetValues<T>() where T : Enum
+        public static T[] GetValues<T>(bool dropObsoleteValues = false) where T : Enum
         {
-            return (T[])Enum.GetValues(typeof(T));
+            var enumValues = (T[])Enum.GetValues(typeof(T));
+
+            if (dropObsoleteValues)
+            {
+                enumValues = enumValues
+                    .Where(v => !IsObsoleteValue(v))
+                    .ToArray();
+            }
+
+            return enumValues;
+        }
+
+        public static bool IsObsoleteValue<T>(T value)
+        {
+            var fi = value.GetType().GetField(value.ToString());
+            var attributes = (ObsoleteAttribute[])fi.GetCustomAttributes(typeof(ObsoleteAttribute), false);
+            return attributes != null && attributes.Length > 0;
```

Comment:
This carries a significant perf cost. 
Do I understand it correctly this functionality is added to deal with obsoletions in `AvatarFallbackType` and `AvatarProvider`?

If so, can we rework the logic in a way where enums don't have obsolete values at all, and in AppSettings we compare strings, e.g. (pseudo code):
```cs
string providerValue = GetString("Appearance.AvatarProvider", "");
foreach (AvatarProvider enumValue in EnumHelper.GetValues<AvatarProvider>())
{
    if (string.Equals(providerValue, enumValue.ToString(), StringComparer.OrdinalIgnoreCase))
        return enumValue;
}

AvatarProvider enumValue = Migrate(providerValue);
Save();
return enumValue;
```

Here the hit will be on reading the value. But since we're reading the value anyway, and the size of the enum is small, all other logic cost would negligible, once the initial migration is done. But we won't have additional costs of additional  metadata inspections and allocating new arrays.

What do you think?

As a side note, this is where source generators could provide significant perf benefits by inlining enum traversals at compile time, thus completely removing runtime metadata traversals. E.g. https://github.com/dotnet/winforms/pull/4043

Reply:
I thought about that too and even briefly started to go into that direction but quickly hit a problem (can't remember exactly how the class was named but I can look it up later) where some kind of config cache stored strongly typed settings which caused a lot of problems with casting.

I can give it another try but that might mean that we can't use `SetEnum` in the setter but use `SetString`.

Reply:
> we can't use `SetEnum` in the setter but use `SetString`.

Why? If a user has an older config, we'd migrate it upon starting an app, and persist the new value, which will be one of the new enum values. Am I missing something?

Reply:
Is there a place where such setting migrations are already handled during startup? I couldn't find anything so tried to fit it into the setting properties themselves which indeed is not a very good spot to put the update mechanism.

Reply:
E.g.
https://github.com/gitextensions/gitextensions/blob/969cf1d766c8ad3e8d14276c07c0ba914cfc8d02/GitCommands/UserRepositoryHistory/LocalRepositoryManager.cs#L187-L194

On line:191 we load data from now-deprecated setting `KeyFavouriteHistory`, and migrate it into the new format.

Reply:
I'm currently rewriting the migration code and removed the obsolete enum values as suggested. Now I obviously have to use GetString to migrate obsolete values and hit the same roadblock as described before that GetString/SetString and GetEnum/SetEnum can't be mixed for a single setting because of the way the internal setting cache is structured.

Here is a quick way to reproduce it, just update the `AppSettings.AvatarProvider` like that:

```c#
public static AvatarProvider AvatarProvider
{
    get
    {
        // Use GetString to fill internal cache with setting value (as type string)
        _ = GetString("Appearance.AvatarProvider", "Default");

        // this line now crashes!!! Because SettingsCache doesn't cache the setting value as string,
        // but a parsed/typed variant of the setting.
        return GetEnum("Appearance.AvatarProvider", AvatarProvider.Default);
    }
    set => SetEnum("Appearance.AvatarProvider", value);
}
```

So I guess this are the options we have:

1. Use `GetString` not only for migration but also inside the properties and parse to enums manually.
  **Pro:** Simple, **Con:** Basically just a workaround for a different problem and might cause confusion why `GetEnum` wasn't used.
2. Fix the `SettingsCache` to support those string ↔ enum conversions.
  **Pro:** Probably the right move because it's the underlying issue, **Con:** Even more unrelated changes in a PR regarding avatar images
3. ~Use `[Obsolete]` enum values and strongly typed values for migration.~
  **Con:** We have already ruled that one out 😅

Reply:
> So I guess this are the options we have:
> 
> 1. Use `GetString` not only for migration but also inside the properties and parse to enums manually.
>    **Pro:** Simple, **Con:** Basically just a workaround for a different problem and might cause confusion why `GetEnum` wasn't used.

We can add a comment explaining this.

> 2. Fix the `SettingsCache` to support those string ↔ enum conversions.
>    **Pro:** Probably the right move because it's the underlying issue, **Con:** Even more unrelated changes in a PR regarding avatar images

...and do this at a later stage. 
@ivangrek was tinkering with the settings infra.

Reply:
✔ done